src/Processors/Transforms/Streaming/WatermarkStamper.cpp

```diff
@@ -361,43 +369,129 @@ void WatermarkStamper::processWatermark(Chunk & chunk)
     }
 
     chunk.setColumns(columns, columns[0]->size());
+}
+
+template <WatermarkStrategy strategy>
+bool WatermarkStamper::applyWatermarkImpl(Int64 event_ts)
```

Comment:
The same code smells happen here as this function is dynamically different for different Watermark strategy and we compile the logic in one function by using templates.